rabbit/rabbittransport.go

```diff
@@ -0,0 +1,426 @@
+	}
+
+	// Used to convey a connection error to the caller (we block until listening has begun)
+	errChan := make(chan error)
+
+	tm.Go(func() error {
+		timeout := time.NewTimer(connectTimeout)
+		defer func() {
+			timeout.Stop()
+			t.listenersM.Lock()
+			defer t.listenersM.Unlock()
+			delete(t.listeners, cn)
+			close(rc)
+			close(errChan)
+			log.Debugf("[Typhon:RabbitTransport] Listener %s stopped", cn)
+		}()
+
+		select {
+		case <-t.tomb.Dying():
+			return nil
+		case <-tm.Dying():
+			return nil
+		case <-timeout.C:
+			errChan <- transport.ErrTimeout
+			return nil
+		case <-t.Ready():
+		}
+
+		deliveryChan, rabbitChannel, err := t.connection().Consume(cn)
+		if err != nil {
+			log.Warnf("[Typhon:RabbitTransport] Failed to consume from %s: %v", cn, err)
+			errChan <- err
+			return nil
+		}
+		defer rabbitChannel.Close()
+		errChan <- nil
+		log.Infof("[Typhon:RabbitTransport] Listening on %s…", cn)
+
+		for {
+			select {
+			case <-t.tomb.Dying():
+				return nil
+
+			case <-tm.Dying():
+				return nil
+
+			case delivery, ok := <-deliveryChan:
+				if !ok {
+					log.Warnf("[Typhon:RabbitTransport] Delivery channel closed; stopping listener %s", cn)
+					return nil
+				}
+				go t.handleReqDelivery(delivery, rc)
+			}
+		}
+	})
+	return <-errChan
+}
+
+func (t *rabbitTransport) logId(delivery amqp.Delivery) string {
+	return fmt.Sprintf("%s[%s]", delivery.RoutingKey, delivery.CorrelationId)
+}
+
+func (t *rabbitTransport) Respond(req message.Request, rsp message.Response) error {
+	headers := rsp.Headers()
+	headers["Content-Encoding"] = "response"
+	headers["Service"] = rsp.Service()
+	headers["Endpoint"] = rsp.Endpoint()
+
+	timeout := time.NewTimer(respondTimeout)
+	defer timeout.Stop()
+	select {
+	case <-t.Ready():
+		timeout.Stop()
+	case <-t.tomb.Dying():
+		return tomb.ErrDying
+	case <-timeout.C:
+		return transport.ErrTimeout
+	}
+
+	return t.connection().Publish("", req.Headers()["X-Rabbit-ReplyTo"], amqp.Publishing{
+		CorrelationId: rsp.Id(),
+		Timestamp:     time.Now().UTC(),
+		Body:          rsp.Payload(),
+		Headers:       headersToTable(headers),
+	})
+}
+
+func (t *rabbitTransport) Send(req message.Request, _timeout time.Duration) (message.Response, error) {
+	id := req.Id()
+	if id == "" {
+		_uuid, err := uuid.NewV4()
+		if err != nil {
+			log.Errorf("[Typhon:RabbitTransport] Failed to generate request uuid: %v", err)
+			return nil, err
+		}
+		req.SetId(_uuid.String())
+	}
+
+	rspQueue := req.Id()
+	defer func() {
+		t.inflightReqsM.Lock()
+		delete(t.inflightReqs, rspQueue)
+		t.inflightReqsM.Unlock()
+	}()
+	rspChan := make(chan message.Response, 1)
+	t.inflightReqsM.Lock()
+	t.inflightReqs[rspQueue] = rspChan
+	t.inflightReqsM.Unlock()
+
+	timeout := time.NewTimer(_timeout)
+	defer timeout.Stop()
+
+	headers := req.Headers()
+	headers["Content-Encoding"] = "request"
+	headers["Service"] = req.Service()
+	headers["Endpoint"] = req.Endpoint()
+
+	select {
+	case <-t.Ready():
+	case <-timeout.C:
```

Comment:
We're using the same timeout here for both connection and receiving a response, so this is treated more as a deadline, and could lead to us sending requests and immediately aborting listening for the responses as a result. Again, fine for now as this is a vast improvement!